lib/Net/SAML2/IdP.pm

```diff
@@ -172,12 +172,14 @@ sub new_from_xml {
         $data->{DefaultFormat} = 'unspecified' unless exists $data->{DefaultFormat};
     }
 
-    for my $key (
-        $xpath->findnodes('//md:EntityDescriptor/md:IDPSSODescriptor/md:KeyDescriptor'))
-    {
-        my $use = $key->getAttribute('use') || 'signing';
+    my @certs = ();
+    my $key_nodeset =
+        $xpath->findnodes('//md:EntityDescriptor/md:IDPSSODescriptor/md:KeyDescriptor');
 
+    while (my $key = $key_nodeset->shift())
+    {
         $key->setNamespace('http://www.w3.org/2000/09/xmldsig#', 'ds');
+        my $use = $key->getAttribute('use') || 'signing';
```

Comment:
I would say:

* If there is only 1 cert without a 'use' attribute we flag it as both a signing and encryption cert.
* If there isn't a 'use' attribute with more than 1 key we die because we don't know the intention of the other party.